Diabetes Prediction using KNN 



In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
df=pd.read_csv("/content/diabetes (1).csv")

In [3]:
df.head()

Pregnancies  Glucose  BloodPressure  ...  DiabetesPedigreeFunction  Age  Outcome
0            6      148             72  ...                     0.627   50        1
1            1       85             66  ...                     0.351   31        0
2            8      183             64  ...                     0.672   32        1
3            1       89             66  ...                     0.167   21        0
4            0      137             40  ...                     2.288   33        1

[5 rows x 9 columns]

In [4]:
df.describe

<bound method NDFrame.describe of      Pregnancies  Glucose  ...  Age  Outcome
0              6      148  ...   50        1
1              1       85  ...   31        0
2              8      183  ...   32        1
3              1       89  ...   21        0
4              0      137  ...   33        1
..           ...      ...  ...  ...      ...
763           10      101  ...   63        0
764            2      122  ...   27        0
765            5      121  ...   30        0
766            1      126  ...   47        1
767            1       93  ...   23        0

[768 rows x 9 columns]>

Since there are many values which have 0 and might be a hindrance while training the set, we change those values to the mean value of the column so as to get better results.

In [5]:

zero_not_accepted = ['Glucose', 'BloodPressure', 'SkinThickness', 'BMI', 'Insulin']

for column in zero_not_accepted:
    df[column] = df[column].replace(0, np.NaN)
    mean = int(df[column].mean(skipna=True))
    df[column] = df[column].replace(np.NaN, mean)



Defining X(dependent variables) and Y (target label) for the model.

In [6]:
x=df.iloc[:,0:8]

In [7]:
x.head()

Pregnancies  Glucose  BloodPressure  ...   BMI  DiabetesPedigreeFunction  Age
0            6    148.0           72.0  ...  33.6                     0.627   50
1            1     85.0           66.0  ...  26.6                     0.351   31
2            8    183.0           64.0  ...  23.3                     0.672   32
3            1     89.0           66.0  ...  28.1                     0.167   21
4            0    137.0           40.0  ...  43.1                     2.288   33

[5 rows x 8 columns]

In [8]:
y=df.iloc[:,-1]

In [9]:
y.head()

0    1
1    0
2    1
3    0
4    1
Name: Outcome, dtype: int64

Here, we split the data into train and test sets. We use Train data for 
building the model and the test data for checking how accurate the trained model is.

In [10]:
from sklearn.model_selection import train_test_split

In [11]:
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.2, random_state=0)

Standardization of a dataset is a common requirement for many machine learning estimators. So, here we scale the train and test data.

In [12]:
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()

In [13]:
x_train=sc.fit_transform(x_train)
x_test=sc.fit_transform(x_test)

The most important part of the model is here, ie, using the KNN algorithm to get predictions/outcomes from our data.

In [14]:
from sklearn.neighbors import KNeighborsClassifier

In [15]:
knn=KNeighborsClassifier(n_neighbors=10, metric='euclidean')

In [16]:
knn.fit(x_train,y_train)


KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='euclidean',
                     metric_params=None, n_jobs=None, n_neighbors=10, p=2,
                     weights='uniform')

Here, we get the outcomes of the test set, which are stored in "pred_y"

In [17]:
pred_y=knn.predict(x_test)

In [18]:
pred_y

array([1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1,
       1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [19]:
y_test

661    1
122    0
113    0
14     1
529    0
      ..
476    1
482    0
230    1
527    0
380    0
Name: Outcome, Length: 154, dtype: int64

Here, we check the accuracy of the model we built.

In [20]:
from sklearn.metrics import accuracy_score

In [21]:
accuracy_score(y_test,pred_y)


0.7922077922077922

In [22]:
from sklearn.metrics import confusion_matrix


In [23]:
confusion_matrix(y_test,pred_y)

array([[97, 10],
       [22, 25]])

Here, we predict the outcome based on values given as input from our end, i.e, we give values for Pregnancies, Glucose, BloodPressure,SkinThickness, Insulin, BMI, DiabetesPedigreeFunction,Age	and get the Outcome.

In [24]:
knn.predict([[1,85	,66	,29	,94	,26.6,	0.351	,31]])

array([1])

In [25]:
knn.predict([[0,137,40,35,168,43.1,2.288,331]])

array([1])

Hence, our model is built, with an accuracy of 79.2%

Now, we use Logistic Regression method for the same dataset to see how it works.

In [26]:
from sklearn.linear_model import LogisticRegression


In [27]:
lr = LogisticRegression()

In [28]:
lr.fit(x_train,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [29]:
y_predlr=lr.predict(x_test)

In [30]:
y_predlr

array([1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1,
       1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0])

In [35]:
#checking predicted value by giving inputs for parameters
lr.predict([[0,137,40,35,168,43.1,2.288,331]])

array([1])

Determining Accuracy score of our Logistic Regression model.

In [32]:
accuracy_score(y_test,y_predlr)

0.8181818181818182

In [34]:
confusion_matrix(y_test,y_predlr)

array([[97, 10],
       [18, 29]])

Thus, we built a Logistic Regression model with 81.8% accuracy.